### PR TITLE
Add support for querying the negotiated TLS version.  The Quickening

### PR DESCRIPTION
**Pull request by elitest**
This PR hopefully finished up the work done by @richmoore in #184.  Thanks to @alex for helping me figure out how to get SSL_get_version implemented.  This PR requires the merge that happened in pyca/cryptography#1869, so do we wait until the next release of cryptography and then bump the required version number in setup.py?

I'm happy to hear feedback on whether the change to strings requires better/different testing?

Also the [OpenSSL docs](https://www.openssl.org/docs/ssl/SSL_get_version.html) have been updated since the original PR in January, and this does actually support TLS1.1 and TLS1.2 :)

### CODE REVIEW

File: doc/api/ssl.rst [hunk 1]
```diff
@@ -598,6 +598,11 @@ Connection objects have the following methods:
     but not it returns the entire list in one go.
 
 
+.. py:method:: Connection.get_protocol_version()
+
+    Retrieve the version of the SSL or TLS protocol used by the Connection
```
Comment (alex) [inline doc/api/ssl.rst]: This should include an example of what the result looks like. Also should end with a `.` :-)

File: OpenSSL/test/test_ssl.py [hunk 2]
```diff
@@ -2745,6 +2745,20 @@ def test_get_cipher_bits(self):
         self.assertEqual(server_cipher_bits, client_cipher_bits)
 
 
+    def test_get_protocol_version(self):
+        """
+        :py:obj:`Connection.get_protocol_version()` returns a string
+        giving the protocol version of the current connection.
+        """
+        server, client = self._loopback()
+        server_protocol_version, client_protocol_version = \
```
Comment (hynek) [inline OpenSSL/test/test_ssl.py]: Please split this into two proper lines.  The current version takes the same amount of lines and is less readable.

File: OpenSSL/SSL.py [hunk 3]
```diff
@@ -1883,6 +1883,18 @@ def get_cipher_version(self):
...
+    def get_protocol_version(self):
+        """
+        Obtain the protocol version of the current connection.
+
+        :returns: The TLS version of the current connection, for example
+            the value for TLS 1.2 would be 0x303.
+        :rtype: :py:class:`unicode`
+        """
+        version = _ffi.string(_lib.SSL_get_version(self._ssl))
+        return version.decode("utf-8")
```
Comment (hynek) [inline OpenSSL/SSL.py]: I’m not sure about this one.  I believe we decided to keep pyOpenSSL’s APIs bytes-based except for paths.

So unless someone can enlighten me on what I’m missing, I’m gonna claim this ought to be bytes.

Comment (alex) [inline OpenSSL/SSL.py]: Dunno, the name of the protocol version definitely seems textual to me.

On Mon, Apr 27, 2015 at 10:16 AM, Hynek Schlawack <notifications@github.com>
wrote:

> In OpenSSL/SSL.py
> <https://github.com/pyca/pyopenssl/pull/244#discussion_r29149506>:
>
> > @@ -1883,6 +1883,18 @@ def get_cipher_version(self):
> >              return version.decode("utf-8")
> >
> >
> > +    def get_protocol_version(self):
> > +        """
> > +        Obtain the protocol version of the current connection.
> > +
> > +        :returns: The TLS version of the current connection, for example
> > +            the value for TLS 1.2 would be 0x303.
> > +        :rtype: :py:class:`unicode`
> > +        """
> > +        version = _ffi.string(_lib.SSL_get_version(self._ssl))
> > +        return version.decode("utf-8")
>
> I’m not sure about this one. I believe we decided to keep pyOpenSSL’s APIs
> bytes-based except for paths.
>
> So unless someone can enlighten me on what I’m missing, I’m gonna claim
> this ought to be bytes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/pyopenssl/pull/244/files#r29149506>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (elitest) [inline OpenSSL/SSL.py]: If I understand your argument correctly, you are saying that when you query for TLS version you should receive something like 0x769 as a response?  The whole python3 bytes v. unicode thing is still new to me...  What was the thinking behind trying to stick with bytes?

Plenty of things in SSL.py are unicode strings though, including much of the related cipher information.  Should a developer really have to get out their OpenSSL decoder ring everytime she wants to query a cipher?  I think ``TLS_RSA_WITH_AES_256_CBC_SHA256`` makes a lot more sense than ``0x3d``, as long as the upstream doesn't change the formatting of those strings.

File: OpenSSL/test/test_ssl.py (hunk 2 above)
Comment (elitest) [inline OpenSSL/test/test_ssl.py]: Excellent point and done in 50d7d15.

File: OpenSSL/SSL.py (hunk 3 above)
Comment (hynek) [inline OpenSSL/SSL.py]: For the first paragraph; what I’m saying is that the docstring is plain wrong.  There’s no way you get anything resembling “0x303” back.  You get a string like `"TLSv1"` back.  Try it and pdb in your test. :)

Comment (elitest) [inline OpenSSL/SSL.py]: My apologies, I should have looked at the code you were commenting on more closely!  I thought I had updated it from the way it was initially proposed in #184.  Fixed in 6ac2e2a.

**Comment by elitest** [comment]
When does Travis update the version of cryptography it tests against?  Based on the way I read the .travis.yml, Travis is supposed to pull from cryptography's master branch?  If that is the case why would the test fail?

**Comment by hynek** [comment]
Good question.  @reaperhulk any ideas?

**Comment by reaperhulk** [comment]
When we switched to using tox in travis we didn't change it over to using the master branch properly, so it's installing it and then not using it.

**Comment by elitest** [comment]
Is that something we want to fix?  If I understand [.travis.yml#L71-83](https://github.com/pyca/pyopenssl/blob/master/.travis.yml#L71-83) properly the problem is that we aren't installing cryptography's master into the virtualenv?  So when it installs pyOpenSSL into the virtualenv via setup.py it installs the latest released cryptography from PyPI?

**Comment by reaperhulk** [comment]
Yeah that needs to be fixed. Right now it installs it into a venv, but then invokes tox (which builds its own venv and doesn't use the master branch).

**Comment by hynek** [comment]
Please rebase on master; this issue should be fixed now.

**Comment by hynek** [comment]
Also, until cryptography 0.9 is out, we can bikeshed on the return type.

My impression is, that pyOpenSSL is bytes only.  If you’d like to argue for unicode I would like to ask you for precedents.

pyOpenSSL is supposed to by a low-level layer around OpenSSL so I’m very apprehensive to introduce unicode-based APIs.  We’ve just added a ton of warning if certain methods are called with unicode.

**Comment by elitest** [comment]
@hynek have you had a chance to look at the [original PR](https://github.com/pyca/pyopenssl/pull/184)?  Would you prefer that instead?

**Comment by hynek** [comment]
Oi, that's where the incorrect doc string came from. :)

That's not what I meant (I just want to get rid of the decode call) but we should definitely expose both ways. I'm on my phone for the next hours but I'll look into it when I'm back home.

It'll be a different PR though. so don't worry about it here.

**Comment by elitest** [comment]
So I might have some time to work on this during the coming weekend in.  Just to be clear, when you say we should expose both ways you mean exposing both ```SSL_get_version```(a.k.a. "TLV1.2" in bytes) and ```SSL_version```(a.k.a. 0x303)?  In #184 it was discussed how to make it clear what is being returned to you, in the documentation, do you have any suggestions there?

**Comment by hynek** [comment]
So I’m saying we should expose both but we start with the string-based version.  And I’m also saying it should return bytes (not unicode) unless you can make a good case for unicode based on existing APIs. :)

**Comment by hynek** [comment]
Can you rebase please?

**Comment by elitest** [comment]
**WIP**
I'm still having issues wrapping my mind around the whole bytes thing... So tests currently fail.  Right now I'm getting stuff like:

>AssertionError: \<cdata 'char *' 0x7fbe700c6c96\> is not an instance of \<class 'bytes'\>

do I need to wrap the result that I'm returning in ```bytes()``` or something?

**Comment by reaperhulk** [comment]
`SSL_get_version` returns a `const char *`. cffi represents that as a proxy object. If you want to obtain the value you'll need to use `ffi.buffer()` or `ffi.string()`. The former requires you either know the length of the `char *` or own all the bytes while the latter reads until it encounters a null (as it assumes a null terminated string). In this case `ffi.string` is the appropriate method

**Comment by elitest** [comment]
That is what I had in whatever 41ac0c4 was before the rebase.  Are we saying that we just don't decode it to utf-8 when we return?

**Comment by elitest** [comment]
Tests should now pass.

**Comment by reaperhulk** [comment]
needs a rebase for travis to test it

**Comment by reaperhulk** [comment]
@hynek this LGTM but since you had some API opinions around bytes earlier I'll let you merge if you think it's ready :)

File: doc/api/ssl.rst [hunk 4]
```diff
@@ -598,6 +598,22 @@ Connection objects have the following methods:
     but not it returns the entire list in one go.
 
 
+.. py:method:: Connection.get_protocol_version()
+
+    Retrieve the version of the SSL or TLS protocol used by the Connection.
+    For example, it will return ``0x303`` for connections made over TLS
+    version 1.2, or ``Unknown`` for connections that were not successfully
```
Comment (hynek) [inline doc/api/ssl.rst]: that seems unlikely :)

File: OpenSSL/SSL.py [hunk 5]
```diff
@@ -1883,6 +1883,18 @@ def get_cipher_version(self):
             return version.decode("utf-8")
 
 
+    def get_protocol_version_name(self):
+        """
+        Obtain the protocol version of the current connection.
+
+        :returns: The TLS version of the current connection, for example
+            the value for TLS 1.2 would be ``b'TLSv1.2'``.
+        :rtype: :py:class:`unicode`
```
Comment (hynek) [inline OpenSSL/SSL.py]: I believe that’s not true anymore?

File: OpenSSL/SSL.py [hunk 6]
```diff
@@ -1938,6 +1950,17 @@ def get_alpn_proto_negotiated(self):
         return _ffi.buffer(data[0], data_len[0])[:]
 
 
+    def get_protocol_version(self):
```
Comment (hynek) [inline OpenSSL/SSL.py]: please group those two methods together

File: ChangeLog [hunk 7]
```diff
@@ -1,3 +1,11 @@
+2015-05-27  Jim Shaver <dcypherd@gmail.com>
+
+	* OpenSSL/SSL.py, : Add ``get_protocol_version()`` and
```
Comment (hynek) [inline ChangeLog]: extra comma plus space

File: OpenSSL/SSL.py [hunk 8]
```diff
@@ -1938,6 +1950,17 @@ def get_alpn_proto_negotiated(self):
...
+        """
+        Obtain the protocol version of the current connection.
+
+        :returns: The TLS version of the current connection, for example
+            the value for TLS 1.2 would be 0x303.
+        :rtype: :py:class:`int`
+        """
+        version = _lib.SSL_version(self._ssl)
+        return version
+
```
Comment (hynek) [inline OpenSSL/SSL.py]: this needs three empty lines

File: ChangeLog [hunk 9]
```diff
@@ -1,3 +1,11 @@
+2015-05-27  Jim Shaver <dcypherd@gmail.com>
+
+	* OpenSSL/SSL.py, : Add ``get_protocol_version()`` and 
+	  ``get_protocol_version_name()`` to Connection
+	  Based on work from Rich Moore
+	* OpenSSL/test/test_crypto.py: tests for ``get_protocol_version()``
```
Comment (hynek) [inline ChangeLog]: just add  OpenSSL/test/test_crypto.py to the list above.  it’s implicit that we add tests for new code. :)

File: ChangeLog [hunk 10]
```diff
@@ -1,3 +1,11 @@
+2015-05-27  Jim Shaver <dcypherd@gmail.com>
+
+	* OpenSSL/SSL.py, : Add ``get_protocol_version()`` and 
+	  ``get_protocol_version_name()`` to Connection
```
Comment (hynek) [inline ChangeLog]: I’d like Connection to be in `` too.  Also please add some periods after sentences. :)

**Comment by hynek** [comment]
I didn’t have opinions as much as I’m trying to figure out what the correct way to do is. :)  As far as I can see, bytes is the way to go but if anyone can show me similar APIs that take/return unicode, I’ll change my mind.

**Comment by elitest** [comment]
I'm fine either way.  There is some [Cipher](https://github.com/pyca/pyopenssl/blob/master/OpenSSL/SSL.py#L1839-1883) stuff that is returned in unicode for example.  But if a discussion has happened around it and that is the direction we are taking additions I'm fine with that.

File: doc/api/ssl.rst (hunk 4 above)
Comment (elitest) [inline doc/api/ssl.rst]: removed unknown in 208438c

File: OpenSSL/SSL.py (hunk 5 above)
Comment (elitest) [inline OpenSSL/SSL.py]: 208438c now bytes

File: OpenSSL/SSL.py (hunk 6 above)
Comment (elitest) [inline OpenSSL/SSL.py]: moved in 208438c

File: OpenSSL/SSL.py (hunk 8 above)
Comment (elitest) [inline OpenSSL/SSL.py]: there are three lines at the bottom, it has been moved 208438c

File: ChangeLog (hunk 10 above)
Comment (elitest) [inline ChangeLog]: 208438c

File: ChangeLog (hunk 9 above)
Comment (elitest) [inline ChangeLog]: a923e93

**Comment by hynek** [comment]
Ugghhh, and those methods are actually pretty analogue to these ones. :-/  Same for Curve.name

*sigh*

Unicode it is then. Sorry.

**Comment by elitest** [comment]
LOL!  It is my own fault I should have pointed it out earlier.

File: OpenSSL/SSL.py [hunk 11]
```diff
@@ -1883,6 +1883,31 @@ def get_cipher_version(self):
             return version.decode("utf-8")
 
 
+    def get_protocol_version_name(self):
+        """
+        Obtain the protocol version of the current connection.
+
+        :returns: The TLS version of the current connection, for example
+            the value for TLS 1.2 would be ``TLSv1.2``or ``Unknown``
+            for connections that were not successfully.
```
Comment (hynek) [inline OpenSSL/SSL.py]: successfully what?

File: doc/api/ssl.rst [hunk 12]
```diff
@@ -598,6 +598,20 @@ Connection objects have the following methods:
...
+.. py:method:: Connection.get_protocol_version()
+
+    Retrieve the version of the SSL or TLS protocol used by the Connection.
+    For example, it will return ``0x769`` for connections made over TLS
+    version 1.
+
+
+.. py:method:: Connection.get_protocol_version_name()
+
+    Retrieve the version of the SSL or TLS protocol used by the Connection.
```
Comment (hynek) [inline doc/api/ssl.rst]: “…as an unicode string.” would be nice to differentiate it from the other method, no?

**Comment by hynek** [comment]
Thank you for your work and your patience!

**Comment by elitest** [comment]
Thanks for your guidance!

**Pull request merged**